src/source.rs

```diff
@@ -1,62 +1,70 @@
-use std::borrow::Borrow;
-use std::cell::RefCell;
-use std::collections::HashMap;
-use std::fmt;
-use std::future::Future;
-use std::hash::{Hash, Hasher};
-use std::pin::Pin;
-use std::rc::Rc;
-
-use futures::future::{FutureExt, Shared};
-use unic_langid::LanguageIdentifier;
+use crate::{fluent::FluentResource, StringFetcher};
 
-use crate::fluent::FluentResource;
-use crate::gecko;
+use futures::{future::Shared, Future, FutureExt};
+use std::{
+    borrow::Borrow,
+    cell::RefCell,
+    collections::HashMap,
+    fmt,
+    hash::{Hash, Hasher},
+    pin::Pin,
+    rc::Rc,
+    task::Poll,
+};
+use unic_langid::LanguageIdentifier;
 
 pub type RcResource = Rc<FluentResource>;
-pub type ResourceOption = Option<RcResource>;
+pub type RcResourceOption = Option<RcResource>;
+pub type ResourceFuture = Shared<Pin<Box<dyn Future<Output = RcResourceOption>>>>;
 
 #[derive(Debug, Clone)]
 pub enum ResourceStatus {
-    Value(RcResource),
-    Async(Shared<Pin<Box<dyn Future<Output = ResourceOption>>>>),
-    None,
+    /// The resource is missing.  Don't bother trying to fetch.
+    Missing,
+    /// The resource is loading and future will deliver the result.
+    Loading(ResourceFuture),
+    /// The resource is loaded and parsed.
+    Loaded(RcResource),
 }
 
-impl From<ResourceOption> for ResourceStatus {
-    fn from(input: ResourceOption) -> Self {
+impl From<RcResourceOption> for ResourceStatus {
+    fn from(input: RcResourceOption) -> Self {
         if let Some(res) = input {
-            Self::Value(res)
+            Self::Loaded(res)
         } else {
-            Self::None
+            Self::Missing
         }
     }
 }
 
-async fn read_resource<P: AsRef<str>>(path: P) -> ResourceOption {
-    gecko::fetch(path.as_ref())
-        .await
-        .ok()
-        .map(|source| Rc::new(FluentResource::try_new(source).unwrap()))
-}
+impl Future for ResourceStatus {
+    type Output = RcResourceOption;
 
-fn set_resolved(
-    cache: Rc<RefCell<HashMap<String, ResourceStatus>>>,
-    full_path: String,
-    value: ResourceOption,
-) {
-    cache
-        .try_borrow_mut()
-        .unwrap()
-        .insert(full_path, value.into());
+    fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        use ResourceStatus::*;
+
+        let this = &mut *self;
+
+        match this {
+            Missing => None.into(),
+            Loaded(res) => Some(res.clone()).into(),
+            Loading(res) => Pin::new(res).poll(cx),
+        }
+    }
 }
 
+/// This is an implementation of a Gecko File Source.  It can be used to return
+/// Fluent resources from Omnija, etc.
```

Comment:
I think this comment is misleading. This is a `FileSource` and can be anything. In Gecko we'll provide a `StringFetcher` and a set of `FileSource`'s, but this crate is more generic.